internal/mesh/internal/controllers/routes/generate.go

```diff
@@ -0,0 +1,631 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package routes
+
+import (
+	"github.com/hashicorp/consul/internal/catalog"
+	"github.com/hashicorp/consul/internal/mesh/internal/controllers/routes/loader"
+	"github.com/hashicorp/consul/internal/mesh/internal/types"
+	"github.com/hashicorp/consul/internal/resource"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+// GenerateComputedRoutes walks a set of related resources and assembles them
+// into one effective easy-to-consume ComputedRoutes result.
+//
+// Any status conditions generated during traversal can be queued for
+// persistence using the PendingStatuses map.
+//
+// This should not internally generate, nor return any errors.
+func GenerateComputedRoutes(
+	related *loader.RelatedResources,
+	pending PendingStatuses,
+) []*ComputedRoutesResult {
+	out := make([]*ComputedRoutesResult, 0, len(related.ComputedRoutesList))
+	for _, computedRoutesID := range related.ComputedRoutesList {
+		out = append(out, compile(related, computedRoutesID, pending))
+	}
+	return out
+}
+
+type ComputedRoutesResult struct {
+	ID      *pbresource.ID
+	OwnerID *pbresource.ID
+	// If Data is empty it means delete if exists.
+	Data *pbmesh.ComputedRoutes
+}
+
+func compile(
+	related *loader.RelatedResources,
+	computedRoutesID *pbresource.ID,
+	pending PendingStatuses,
+) *ComputedRoutesResult {
+	// There is one computed routes resource for the entire service (perfect name alignment).
+	//
+	// All ports are embedded within.
+
+	parentServiceID := &pbresource.ID{
+		Type:    catalog.ServiceType,
+		Tenancy: computedRoutesID.Tenancy,
+		Name:    computedRoutesID.Name,
+	}
+
+	parentServiceRef := resource.Reference(parentServiceID, "")
+
+	parentServiceDec := related.GetService(parentServiceID)
+
+	allowedPortProtocols := make(map[string]pbcatalog.Protocol)
+	inMesh := false
+	if parentServiceDec != nil {
+		parentServiceID = parentServiceDec.Resource.Id // get ULID out of it
+
+		for _, port := range parentServiceDec.Data.Ports {
+			if port.Protocol == pbcatalog.Protocol_PROTOCOL_MESH {
+				inMesh = true
+				continue // skip
+			}
+			allowedPortProtocols[port.TargetPort] = port.Protocol
+		}
+	}
```

Comment:
> We should probably add pending status also saying that the referenced service is not a mesh service (that maybe is done later but I wonder why we don't do it here)

That is handled in `ref_validation.go` before this is called.

Reply:
Rewrote to handle the early-exit for those 2 cases.

Reply:
> We should probably add pending status also saying that the referenced service is not a mesh service (that maybe is done later but I wonder why we don't do it here)

We don't do it here, because if an xRoute has multiple parents, it would be part of multiple ComputedRoutes and we'd end up adding the same status condition twice.